### PR TITLE
Fix stack overflow in the configuration source gen

### LINKED ISSUE #103802: StackOverflowException in Configuration.Binder source generator
### Repro steps

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableConfigurationBindingGenerator>true</EnableConfigurationBindingGenerator>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="8.0.0" />
  </ItemGroup>
  
</Project>
```

```csharp
using Microsoft.Extensions.Configuration;

namespace ClassLibrary7;

public class Class1
{
    public static void Bind(IConfiguration config)
    {
        var supported = new ExampleSettings();
        config.Bind(supported);
    }
}

public record ExampleSettings
{
    /// <summary>
    /// A recursive data structure
    /// </summary>
    public TreeElement? Tree { get; set; }
}

/// <summary>
/// Represents a free-format data structure.
/// </summary>
public sealed class TreeElement : Dictionary<string, TreeElement>;
```

### Expected results

I'd expect the code to compile successfully.

### Actual results

This results in a StackOverflowException in the Configuration Binder source generator. When this happens, it locks VS up and you need to force kill it (or let it eventually crash).

_Originally posted by @eerhardt in https://github.com/dotnet/aspire/pull/4441#discussion_r1648262424_

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/103802

We'll need to port this to the rc2/release branch.

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 1]
```diff
@@ -562,9 +564,28 @@ private bool IsUnsupportedType(ITypeSymbol type)
                     return true;
                 }
 
+                if (rootCall)
+                {
+                    ts_visitedTypes ??= new HashSet<ITypeSymbol>(SymbolEqualityComparer.Default);
+                    ts_visitedTypes.Clear();
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:570]: It's odd for this to be thread-static state, then the clearing behavior to be controlled by a parameter only specified by this function.  Could you instead just pass this as the optional parameter, and create it when the value is null?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs [hunk 2]
```diff
@@ -188,6 +188,14 @@ public string Color
             }
         }
 
+        public sealed class TreeElement : Dictionary<string, TreeElement>;
+
+        public record TypeWithRecursionThroughCollections
+        {
+            public TreeElement? Tree { get; set; }
+            public TreeElement?[]? Flat { get; set; }
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.TestClasses.cs:196]: Can you also make sure you hit the dictionary and collection case too?

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs [hunk 3]
```diff
@@ -1544,6 +1544,79 @@ public void CanBindOnParametersAndProperties_RecordWithArrayConstructorParameter
...
+
+            var configuration = new ConfigurationBuilder()
+                        .AddJsonStream(new System.IO.MemoryStream(System.Text.Encoding.UTF8.GetBytes(jsonConfig)))
+                        .Build();
+
+            var instance = new TypeWithRecursionThroughCollections();
+            configuration.Bind(instance);
+
+            Assert.NotNull(instance.Tree);
+            Assert.Equal(2, instance.Tree.Count);
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:1589]: I wouldn't bother being so verbose here.  I think making sure the top level tree count is correct and do one depth assertion to ensure the deepest element is set is enough.

File: src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs [hunk 4]
```diff
@@ -545,7 +545,9 @@ private bool IsAssignableTo(ITypeSymbol source, ITypeSymbol dest)
                 return conversion.IsReference && conversion.IsImplicit;
             }
 
-            private bool IsUnsupportedType(ITypeSymbol type)
+            private static HashSet<ITypeSymbol>? s_visitedTypes;
```
Comment (ericstj) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs:548]: Why is it static?  Can it be instance state and not require interlocked?

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingGenerator.Parser.cs]: I have changed that to be instance field instead.

**Review by ericstj — APPROVED** [review]
LGTM

**Pull request merged**